src/Adapter/ContextStateManager.php

```diff
@@ -218,6 +245,18 @@ private function restoreContextField(string $fieldName): void
         $currentStashIndex = array_key_last($this->contextFieldsStack);
         // NOTE: array_key_exists important here, isset cannot be used because it would not detect if null is stored
         if (array_key_exists($fieldName, $this->contextFieldsStack[$currentStashIndex])) {
+            if ('shop' === $fieldName) {
```

Comment:
So much logic, makes it tempting to even address it in a dedicated function `restoreShopContext()` ?

Reply:
Yeah I also hesitated, I'm gonna add this one